jaeger/spans.go

```diff
@@ -0,0 +1,121 @@
+package jaeger
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"sort"
+	"strconv"
+	"time"
+
+	jaegerModels "github.com/jaegertracing/jaeger/model/json"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+)
+
+// TODO / Question: make limit configurable? Selected from UI?
+const tracesLimit = 100
+
+type Span struct {
+	jaegerModels.Span
+	TraceSize int `json:"traceSize"`
+}
+
+func getSpans(client http.Client, endpoint *url.URL, namespace, service, startMicros, endMicros string) ([]Span, error) {
+	u := endpoint
+	queryService := service
+	if config.Get().ExternalServices.Tracing.NamespaceSelector {
+		queryService = service + "." + namespace
+	}
+	u.Path = path.Join(u.Path, "/api/traces")
+	q := u.Query()
+	q.Set("service", queryService)
+	q.Set("start", startMicros)
+	q.Set("end", endMicros)
+	q.Set("limit", strconv.Itoa(tracesLimit))
+
+	u.RawQuery = q.Encode()
+	traces, _, err := queryTraces(client, u)
+	if err != nil {
+		return []Span{}, err
+	}
+
+	spans := tracesToSpans(traces, service, namespace)
+	if len(traces) == tracesLimit {
+		// Reached the limit, trying to be smart enough to show more and get the most relevant ones
+		log.Info("Limit of traces was reached, trying to find more relevant spans...")
+		return findRelevantSpans(client, spans, u, q, service, namespace)
+	}
+
+	return spans, nil
+}
+
+func tracesToSpans(traces []jaegerModels.Trace, service, namespace string) []Span {
+	spans := []Span{}
+	for _, trace := range traces {
+		// First, get the desired processes for our service
+		processes := make(map[jaegerModels.ProcessID]bool)
+		for pId, process := range trace.Processes {
+			if process.ServiceName == service || process.ServiceName == service+"."+namespace {
+				processes[pId] = true
+			}
+		}
+		// Second, find spans for these processes
+		for _, span := range trace.Spans {
+			if ok := processes[span.ProcessID]; ok {
+				spans = append(spans, Span{
+					Span:      span,
+					TraceSize: len(trace.Spans),
+				})
+			}
+		}
+	}
+	log.Infof("Found %d spans in the %d traces for service %s", len(spans), len(traces), service)
+	return spans
+}
+
+func findRelevantSpans(client http.Client, spansSample []Span, u *url.URL, q url.Values, service, namespace string) ([]Span, error) {
```

Comment:
Nice logic here.
I know it's quite far, but conceptually it opens to add things like this in the future:
https://www.hawkular.org/blog/2017/08/alerts-nelson-rules.html

Reply:
Thanks! This is very interesting in particular in regard of what I'd like to experiment here https://github.com/kiali/kiali/issues/1981 . We should definitely talk about that tomorrow if possible!